[IMP] sale_invoice_policy: Several improvements

rousseldenis opened the pull request:
- [x] Don't use the required parameter to compute the quantity to invoice
- [x] Simplify `depends()`
- [x] Don't use the required parameter to compute the untaxed amount to invoice
- [x] Don't rewrite the Odoo core compute method
- [x] Improve README

Refactor:

I think the default policy that was applied on sale orders from the default product configuration (defined in sale module) was the wrong approach as it is not contextually correct (products are not sale orders).

That said, I introduced an invoice policy on sale order level more explicitly than before (void) which is 'Products Invoicing Policy'. That let the user free to choose between standard behavior or sale order level configuration.

The question is:

- As the required attribute was used before means that the sale order invoicing policy should not use the product configurations. Should we reintroduce that behavior (to constrain sale orders to NOT use product configuration at all ?) ? @ioans73 @fuentes73 @sergio-teruel @ACheung-FactorLibre @Cedric-Pigeon @acsonefho

rousseldenis committed:
[FIX] sale_invoice_policy: Don't use the 'required' parameter

Don't use the 'required' parameter to compute the quantities to invoice or not.
In fact, this should only depends on the invoice policy itself.

rousseldenis committed:
[IMP] sale_invoice_policy: Don't reuse fields in depends

As ORM appends new depends() parameters to existing ones, simplify the writing

rousseldenis committed:
[IMP] sale_invoice_policy: Don't rely on required option to compute the amount

As for the 'qty_to_invoice' field, the 'untaxed_amount_to_invoice' field should not
depends on the required option to compute its value

rousseldenis committed:
[IMP] sale_invoice_policy: Don't rewrite Odoo core compute method

As we don't want to rewrite Odoo core method inside this module, we
should not rewrite the method to compute the untaxed amount to invoice.

Instead, save the product invoice policy, change it to sale order one without
recomputing context, call the super() method and then, restore its former value.

rousseldenis committed:
[IMP] sale_invoice_policy: save products invoice policy values in a separate dict

rousseldenis committed:
[IMP] sale_invoice_policy: Add the default behavior as sale invoice policy

With default behavior in Odoo, the invoicing policy depends on products configuration.
So, include this as possible value for sale order invoicing policy. Possible values are so:

  - Products Invoice Policy: each product applies its invoicing policy (by sale line)
  - Ordered quantities: the whole sale order will apply the invoicing policy on ordered quantities
  - Delivered Quantities: the whole sale order will apply the invoicing policy on delivered quantities

That said, the 'required' configuration is no more useful as the invoicing policy is now required
on sale order level.

rousseldenis commented:
Don't hesitate to test it on runboat

lmignon commented on sale_invoice_policy/models/sale_order_line.py:
```diff
@@ -29,74 +25,27 @@ def _compute_qty_to_invoice(self):
+            with self.env.norecompute():
+                saved_product_values[line.product_id] = line.product_id.invoice_policy
+                line.product_id.invoice_policy = line.order_id.invoice_policy
+            # We should call the super() per line as each sale order can have different policy
```
To avoid 'recomputes' and marking the field as modified (i.e. writing it to DB), you need to change the approach and use the ‘protecting’ method on the env. You could also improve readability and the batch processing by grouping your lines by invoice policy and using a context manager to apply the invoice policy to the products when computing the untaxed amount to invoice.

```python

@contextmanager
def _sale_invoice_policy(self, lines):
    """ Apply the sale invoice policy to the products
   
    This method must be called with lines sharing the same invoice policy
    """
    invoice_policy = lines.mapped("order_id.invoice_policy")
    if len(invoice_policy) > 1:
        raise Exception("The mehtod _sale_invoice_policy must be called with lines sharing the same invoice policy")
    invoice_policy = invoice_policy[0]
    invoice_policy_field = self.env["product.product"]._fields["invoice_policy"]
    products = lines.product_id
    with self.env.protecting([invoice_policy_field], products):
        old_values = {}
        for product in products:
            old_values[product] = product.invoice_policy
            product_id.invoice_policy = invoice_policy
        yield
        for product, invoice_policy in old_values.items():
            product.invoice_policy = invoice_policy

@api.depends("order_id.invoice_policy")
def _compute_untaxed_amount_to_invoice(self) -> None:
    other_lines = self.filtered(
        lambda line: line.product_id.type == "service"
        or line.order_id.invoice_policy == "product"
        or line.order_id.invoice_policy == line.product_id.invoice_policy
        or line.state not in ["sale", "done"]
    )
    super(SaleOrderLine, other_lines)._compute_untaxed_amount_to_invoice()
    for lines in (self - other_lines).partition("invoice_policy").values():
        with self._sale_invoice_policy(lines):
            super(SaleOrderLine, lines)._compute_untaxed_amount_to_invoice()
```

rousseldenis replied:
Done

rousseldenis committed:
[IMP] sale_invoice_policy: Use protecting() instead of deprecated norecompute()

To avoid marking 'invoice_policy' product field as dirty for products temporary value,
use the working context manager 'protecting()' instead of deprecated 'norecompute() one.

Put the whole process in its own contextmanager.

lmignon commented on sale_invoice_policy/models/sale_order_line.py:
```diff
@@ -29,74 +50,18 @@ def _compute_qty_to_invoice(self):
-            line.untaxed_amount_to_invoice = amount_to_invoice
-        return True
+        for lines in (
+            (self - other_lines).partition("product_id.invoice_policy").values()
```
```suggestion
            (self - other_lines).partition("order_id.invoice_policy").values()
```

rousseldenis committed:
[IMP] sale_invoice_policy: Improve README

ferran-S73 reviewed (approved):
Code review OK. As per your question I see no problem in allowing product configuration for sale orders' invoice policy. It adds more flexibility to the module

phschmidt reviewed (approved):
Functionary review

marielejeune reviewed (approved):
LGTM (code + functional review)
Thanks for the work

anajuaristi commented:
> Don't hesitate to test it on runboat

Hello @rousseldenis 
I was trying to test in runboat but I can´t find or see the build for this PR. Could you help me with that please? 

Thank you!!!

jbaudoux commented on sale_invoice_policy/models/sale_order.py:
```diff
@@ -9,41 +9,27 @@ class SaleOrder(models.Model):
         help="Ordered Quantity: Invoice based on the quantity the customer "
         "ordered.\n"
         "Delivered Quantity: Invoiced based on the quantity the vendor "
         "delivered (time or deliveries).",
```
Add that is applies only for stock products

jbaudoux reviewed (commented):
Can you add in the readme that the SO invoice policy affects only stock products. Service products continue to be invoiced based on the product invoice policy.

jbaudoux commented on sale_invoice_policy/__manifest__.py:
```diff
@@ -10,9 +10,10 @@
         "views/res_config_settings_view.xml",
         "views/sale_view.xml",
     ],
+    "pre_init_hook": "pre_init_hook",
```
```suggestion
    "pre_init_hook": "pre_init_hook",
    "external_dependencies": {"python": ["openupgradelib"]},
```

jbaudoux reviewed (commented):
And generate `requirement.txt` as it is not part of it in this repo

```
2025-01-25 16:00:20,521 417 INFO odoo odoo.modules.loading: Loading module sale_invoice_policy (66/67) 
2025-01-25 16:00:20,522 417 CRITICAL odoo odoo.modules.module: Couldn't load module sale_invoice_policy 
2025-01-25 16:00:20,522 417 CRITICAL odoo odoo.modules.module: No module named 'openupgradelib' 
```

marielejeune commented:
@rousseldenis Could you have a look at the last reviews please?

rousseldenis commented:
@jbaudoux Could you update your review ?

hildickethan reviewed (approved):
Looks good!

jbaudoux reviewed (approved):
Can you drop auto-generated files from the commits?

alexey-pelykh reviewed (commented):
Thanks for the refactor Denis, the move to a proper three-valued selection with `"product"` as a default is a much cleaner design than the old `invoice_policy_required` boolean approach. The delegation to Odoo core in `_compute_untaxed_amount_to_invoice` instead of reimplementing the whole calculation is a welcome simplification too.

A few things I noticed:

There is a copy-paste typo in `readme/CONTEXT.rst` (and consequently in the generated README.rst and index.html) -- the two bullet points both say "Invoicing on ordered quantities". The second one should be "Invoicing on delivered quantities".

In `_sale_invoice_policy`, a bare `Exception` is raised when called with mixed policies. Using `ValueError` (or `UserError` if the intent is to surface this to the UI) would be more conventional.

As @jbaudoux mentioned in his last review, the auto-generated files (`README.rst`, `static/description/index.html`) should be dropped from the commits. OCA CI regenerates them.

One note on the context manager pattern: the `env.protecting` + temporary product field patching approach is clever, but it does mean that if an exception occurs between setting the new values and the `yield`, the old values would not be restored. In practice this seems unlikely to cause issues since any such exception would typically roll back the transaction, but worth being aware of.

Otherwise the implementation looks solid -- `pre_init_hook` with `openupgradelib` for the migration, the company-level default, tests covering both policy paths and the context manager guard. Nice work.

rousseldenis commented:
/ocabot merge major
